### PR TITLE
Logger and hardened atomic serial work

### LINKED ISSUE #90: Unable to determine board type
Note: You will also see this message if you have not uploaded firmware to the board on every version (Choose Firmware -> Upload Firmware) first.

There is a serial issue where the initial firmware request message is cut short when sent from the board causing the GUI to to be able to tell what try of board is attached.

For now just reconnect. Pertains to version 2.2 only here.

This was caused by the addition of the GIT sha to the firmware.

Edit: Underlying issues with serial CDC, looking into Zephyr's async

### PR DESCRIPTION
**Pull request by gcmcnutt**
Here I'm doing a few things:
- using mutex vs the irq shut down
- trying to make all serial writes atomic that need to be
- working to get the serialWriteF working concurrently
- also adding the second FW fetch during client connect

this is a review step - I've yet to test the various pieces

Fixes #90

### CODE REVIEW

File: firmware/src/src/include/ble.h [hunk 1]
```diff
@@ -43,4 +43,4 @@ void leparamupdated(struct bt_conn *conn,
                            uint16_t timeout);
 void securitychanged(struct bt_conn *conn, bt_security_t level, enum bt_security_err err);
 void lephyupdated(struct bt_conn *conn, struct bt_conn_le_phy_info *param);
-void printPhy(int);
\ No newline at end of file
+const char *printPhy(int);
```
Comment (gcmcnutt) [inline firmware/src/src/include/ble.h:46]: fix this newline

**Comment by dlktdr** [comment]
Took a quick look. You got printf working!! Awesome. From what I see it looks good. I'm back home tomorrow and can give it a try. If just sending FW twice works for now that seems like a workable solution. We can always work on async in next release if possible.

I'm thinking in the next PR after this is merged we remove all the HT: from the logs, and incorporate some base log defines e.g. LOG(int Level, "SRC", "Message", ... ) Now that printf is working. There was really no need for the HT: in all the items. Can have the define push it as Inf, wrn, Dbg, err instead and color code them in the GUI. Have debug outputs automatically disabled on release compile too.

Excellent work!

**Review by dlktdr — APPROVED** [review]
Got a chance to briefly test it tonight. I did get a unable to connect to board. It didn't appear the GUI send out the second fw request. Looks like the fw was in the middle of sending some data back to the GUI. The board did send the full fw command so it might be a GUI issue here. Although I couldn't repeat the issue it worked every time after.

Overall I love the serialwritef functionality is working and the logging functionality works great.

If your happy too, go ahead and merge it. 👍

**Comment by gcmcnutt** [comment]
let me first add in the logger support for hex dump -- at that point can simplify down the serialWrite functions to a minimal set -- and a couple of other TODOs in there --

**Comment by gcmcnutt** [comment]
@dlktdr this might be mergeable -- would be good to see some output from the bytesToHex function -- perhaps you can try a test of this?

Also not sure how to build the ESP32 stuff, so maybe make sure that still works (as I did change some common code).  I've got the hardware and debuggers now if I need to dig into this -- let me know

**Comment by dlktdr** [comment]
For sure, will give it a try,

ESP will not compile at the moment. Not much of the functionality is moved over yet. Hasn't been high on my priority list until the ICM-20948's I ordered are closer to delivery. Hopefully close to Oct but who knows at this point. So no worries about seeing if that will work.

**Pull request merged**